### DISCUSSION TITLE
NotImplementedError on bulk insert setup_dml_returning_compile_state

### DISCUSSION THREAD

**Discussion by MPuust** [sqlalchemy/sqlalchemy#10075] [Usage Questions] [answered]
Using SQLAlchemy==2.0.17

I have a code with which I try to insert bulk rows. Unfortunately this results in NotImplemented error, as this function is not implemented. Is there a workaround for this?
```#db: sqlalchemy.orm.session
#insert: sqlalchemy.dialects.postgresql.dml

insert_statement = insert(self.model).values(obj_in).on_conflict_do_nothing(
            index_elements=unique_fields
        ).returning("*")
rows = db.execute(insert_statement).all()
db.commit()

Error:
File "/venv/lib/python3.8/site-packages/sqlalchemy/orm/context.py", line 2489, in setup_dml_returning_compile_state
    raise NotImplementedError()
Function:
def setup_dml_returning_compile_state(
        self,
        compile_state: ORMCompileState,
        adapter: DMLReturningColFilter,
    ) -> None:
        raise NotImplementedError()

**Comment by zzzeek** [comment]
upgrade to SQLAlchemy 2.0 where all those use cases [are supported](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#using-returning-with-upsert-statements)

if on 1.4, [use the older workaround approach](https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#using-postgresql-on-conflict-with-returning-to-return-upserted-orm-objects)

**Reply by MPuust** [reply]
On SQLAlchemy 2.0.17, added it.

**Reply by MPuust** [reply]
https://github.com/sqlalchemy/sqlalchemy/blob/400aa8a676ba1a0a1536ae52a20caa93726525dd/lib/sqlalchemy/orm/context.py#L2484C37-L2484C37

**Reply by zzzeek** [reply, marked as answer]
dont' return '*', use the entity you want to return.  follow the documentation example

**Reply by MPuust** [reply]
Worked like charm, thanks @zzzeek !
